File: Frontend/lib/Objects/course.dart
class Course {
  String id;
  String term;
  String syllabus;
  String department;
  String coordinatorID;
  String credits;
  List<String> sections;
  List<String> exams;
  Map<String, String> finalGrades;

  Course({
    required this.id,
    required this.term,
    required this.syllabus,
    required this.department,
    required this.coordinatorID,
    required this.credits,
    required this.sections,
    required this.exams,
    required this.finalGrades,
  });

  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      id: json['id'].toString(),
      term: json['term'].toString(),
      syllabus: json['syllabus'].toString(),
      department: json['department'].toString(),
      coordinatorID: json['coordinatorID'].toString(),
      credits: json['credits'].toString(),
      sections: List<String>.from(json['sections'] ?? []),
      exams: List<String>.from(json['exams'] ?? []),
      finalGrades: Map<String, String>.from(json['finalgrades'] ?? {}),
    );
  }
}

File: Backend/models/course.js
const mongoose = require("mongoose");
const { v4: uuidv4 } = require("uuid");
const exam = require("./exam");

const courseSchema = new mongoose.Schema(
    {
        _id: {
            type: String,
            default: () => uuidv4().replace(/\-/g, ""),
          },
          id: String,
          term: String,
          syllabus: String,
          department: String,
          coordinatorID: String,
          credits: String,
          sections: [String],
          exams: [String],
          finalgrades: {String: String}
    },
    {
        timestamps: true,
        collection: "courses",
    }
);

courseSchema.statics.createCourse = async function (id, term, department, coordinatorID, credits) {
    try {
        const course = await this.create({id, term, department, coordinatorID, credits});
        return course;
    } catch (error) {
        throw error;
    }
}

courseSchema.statics.getCourseByID = async function (id) {
    try {
        const course = await this.find({ id: id});
        if(!course) throw ({error: 'No course with this id found' });
        return course;
    } catch (error) {
        throw error;
    }
}

courseSchema.statics.getCourseByTerm = async function (term) {
    try {
        const course = await this.find({ term: term});
        if(!course) throw ({error: 'No course with this term found' });
        return course;
    } catch (error) {
        throw error;
    }
}

courseSchema.statics.getCourseByIDAndTerm = async function (id, term) {
    try {
        const course = await this.find({ id: id, term: term});
        if(!course) throw ({error: 'No course with this id and term found' });
        return course;
    } catch (error) {
        throw error;
    }
}

courseSchema.statics.getCourses = async function () {
    try {
      const courses = await this.find();
      return courses;
    } catch (error) {
      throw error;
    }
}

courseSchema.statics.deleteCourseByIDAndTerm = async function (id, term) {
    try {
      const result = await this.deleteOne({ id: id , term: term});
      return result;
    } catch (error) {
      throw error;
    }
}

courseSchema.statics.editCourseByIDAndTerm = async function (id, term, syllabus, department, coordinatorID, credits, sections, exams, finalgrades) {
    try {
        const course = await this.findOne({id: id, term: term});

        const courseUpdates = {
            syllabus: syllabus,
            department: department,
            coordinatorID: coordinatorID,
            credits: credits,
            sections: sections,
            exams: exams,
            finalgrades: finalgrades
        };

        const courseUpdated = await this.findOneAndUpdate(
            { id: id , term: term},
            { $set: courseUpdates },
            { new: true } 
        );

        console.log("ah");
        return courseUpdated;
    } catch (error) {
        throw error;
    }
}

module.exports = mongoose.model("Course", courseSchema);